source/particle/world.cc

```diff
@@ -833,21 +924,31 @@ namespace aspect
               if (cell->is_locally_owned())
                 moved_particles_cell.push_back(cell_and_particle);
               else
-                moved_particles_domain.insert(std::make_pair(cell->subdomain_id(),cell_and_particle));
+                moved_particles_domain[subdomain_to_neighbor_map[cell->subdomain_id()]].push_back(cell_and_particle);
             }
         }
     }
 
     template <int dim>
     void
-    World<dim>::send_recv_particles(const std::multimap<types::subdomain_id, std::pair<types::LevelInd,Particle <dim> > > &send_particles,
-                                    std::vector<std::pair<types::LevelInd, Particle<dim> > >                              &received_particles)
+    World<dim>::send_recv_particles(const std::vector<std::vector<std::pair<types::LevelInd,Particle <dim> > > > &send_particles,
+                                    std::vector<std::pair<types::LevelInd, Particle<dim> > >                     &received_particles)
     {
       // Determine the communication pattern
       const std::vector<types::subdomain_id> neighbors (this->get_triangulation().ghost_owners().begin(),
                                                         this->get_triangulation().ghost_owners().end());
       const unsigned int n_neighbors = neighbors.size();
 
+      Assert(n_neighbors == send_particles.size(),
+             ExcMessage("The particles to send to other processes should be sorted into a vector "
+                        "containing as many vectors of particles as there are neighbor processes. This "
+                        "is not the case for an unknown reason. Contact the developers if you encounter "
+                        "this error."));
```

Comment:
Not wrong. We commonly use `ExcInternalError()` for this, but not worth changing now.